noir-r1cs/src/sparse_matrix.rs

```diff
@@ -175,10 +202,44 @@ impl Mul<HydratedSparseMatrix<'_>> for &[FieldElement] {
             rhs.matrix.num_rows,
             "Vector length does not match number of rows."
         );
+
+        let intermediate_multiplication =
+            LockFreeArray::new(vec![(0, FieldElement::zero()); rhs.matrix.num_entries()]);
```

Comment:
It's unfortunate that it requires allocating a temporary vector larger than the output.